evergreen-ci/spruce
# EVG-19094: Add error page metadata in Bugsnag

EVG-19094

### Description
(Context: I goofed up during the implementation of [EVG-18872](https://jira.mongodb.org/browse/EVG-18872) and didn't realize I was editing the local error boundary 🫠)

We need to log information about whether the error page has been viewed in Bugsnag. I tried a few different things (listed in more detail below) but the most straightforward way to do this was to add a field in another tab. This field will only show up if the user saw the error page.

<img width="708" alt="Screenshot 2023-03-03 at 3 53 40 PM" src="https://user-images.githubusercontent.com/47064971/222827556-b6f71c1c-b4c8-42a5-8ef6-9ecea51c1f2b.png">

### Steps to manually test
1. Introduce some artificial error into the codebase on any page. For example you can try console logging `some_empty_array[3]`
2. Deploy to staging
3. Turn off your adblock (if you have any)
4. Load the page with the artificial error
5. Go check Bugsnag to view the reported error


###  Other things I tried
* Using `Bugsnag.start`'s [`onError`](https://docs.bugsnag.com/platforms/javascript/customizing-error-reports/#updating-events-using-callbacks) function. 
  * This function runs on any error, even the ones we manually report in the codebase. 
  * Bugsnag exposes some information about if an error is unhandled or handled. Errors that we manually report are handled. However this doesn't necessarily mean that every unhandled error sends the user to the error page. For example, resize observer errors are unhandled but they don't send the user to the error page. Therefore I don't think we can actually tell if the user sees the error page through this function.
* Amending the breadcrumbs.
  * I tried modifying [`event.breadcrumbs`](https://docs.bugsnag.com/platforms/javascript/customizing-error-reports/#breadcrumbs) to see if I could insert an extra breadcrumb about the error, but I could never get it to show up in Bugsnag.
  * Also tried using [`leaveBreadcrumb`](https://docs.bugsnag.com/platforms/javascript/customizing-breadcrumbs/#adding-manual-breadcrumbs), which didn't work either.
* Amending the error class.
  * I tried overwriting [`errors[0].errorClass`](https://docs.bugsnag.com/platforms/javascript/customizing-error-reports/#breadcrumbs) to say something else, which works, but it obscures the original error. If you edit the error class it will show up like below:
  * <img width="637" alt="Screenshot 2023-03-03 at 4 16 03 PM" src="https://user-images.githubusercontent.com/47064971/222830890-b7d8b03a-60be-405c-97fc-dd0a7d445f1c.png">
